File: src/main/java/me/sargunvohra/svlib/capability/PlayerCapability.java
package me.sargunvohra.svlib.capability;

import java.util.HashSet;
import java.util.Set;
import lombok.val;
import net.minecraft.entity.player.EntityPlayer;

public abstract class PlayerCapability {
  private Set<Runnable> listeners = new HashSet<>();

  public void apply(EntityPlayer target) {}

  public boolean shouldPersistOnDeath() {
    return false;
  }

  public final void addListener(Runnable listener) {
    listeners.add(listener);
  }

  public final void notifyModified() {
    for (val listener : listeners) listener.run();
  }
}

File: src/main/java/me/sargunvohra/svlib/capability/PlayerCapability.kt
package me.sargunvohra.svlib.capability

import net.minecraft.entity.player.EntityPlayer
import java.util.HashSet

abstract class PlayerCapability {
    private val listeners = HashSet<() -> Any>()

    open fun apply(target: EntityPlayer) {}

    open fun shouldPersist(wasDeath: Boolean) = !wasDeath

    fun addListener(listener: () -> Unit) = listeners.add(listener)

    fun notifyModified() = listeners.forEach { it() }
}
